### PR TITLE
Stream request body instead of buffering it in memory

### PR DESCRIPTION
**Pull request by hbgl**
## Changes

- Resolves https://github.com/withastro/astro/issues/7525.
- Stream request body instead of buffering it in memory.
- Relies on non-standard behavior in [undici](https://github.com/nodejs/undici#requestbody) that accepts AsyncIterables as request bodies. Node uses undici under the hood, so this is (unofficially) supported.
- Refactor header conversion to make it type-safe.

## Testing

- Existing test cases for sending request bodies.
- Added another test case for posting large request bodies (256MB).

## Docs

- Internal bugfix. The change does not affect user's behavior.

### CODE REVIEW

File: packages/astro/src/core/app/node.ts [hunk 1]
```diff
@@ -9,84 +9,117 @@ import { App, type MatchOptions } from './index.js';
...
+
+	// Return default body.
+	return asyncIterableToBodyProps(req);
+}
+
+function asyncIterableToBodyProps(iterable: AsyncIterable<any>): BodyProps {
+	return {
+		// Node uses undici for the Request implementation. Undici accepts
+		// a non-standard async iterable for the body.
+		// @ts-ignore
```
Comment (ematipico) [inline packages/astro/src/core/app/node.ts:91]: `@ts-expect-error` is a better fit

File: packages/astro/src/core/app/node.ts [hunk 2]
```diff
@@ -9,84 +9,117 @@ import { App, type MatchOptions } from './index.js';
...
+		body: iterable,
+		// The duplex property is required when using a ReadableStream or async
+		// iterable for the body. The type definitions do not include the duplex
+		// property because they are not up-to-date.
+		// @ts-ignore
+		duplex: 'half',
```
Comment (ematipico) [inline packages/astro/src/core/app/node.ts:97]: `@ts-expect-error` is a better fit

File: packages/astro/src/core/app/node.ts [hunk 3]
```diff
@@ -9,84 +9,117 @@ import { App, type MatchOptions } from './index.js';
...
+
+function asyncIterableToBodyProps(iterable: AsyncIterable<any>): BodyProps {
+	return {
+		// Node uses undici for the Request implementation. Undici accepts
+		// a non-standard async iterable for the body.
+		// @ts-ignore
+		body: iterable,
+		// The duplex property is required when using a ReadableStream or async
+		// iterable for the body. The type definitions do not include the duplex
+		// property because they are not up-to-date.
```
Comment (ematipico) [inline packages/astro/src/core/app/node.ts:95]: Is there an updated version that matches the types?

Comment (hbgl) [reply packages/astro/src/core/app/node.ts:95]: There is sadly no updated version with the correct types. There exists an issue https://github.com/DefinitelyTyped/DefinitelyTyped/issues/60924 but no fix yet (see [RequestInit](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f701ddcb5bec7d89238b8f1b9d362a20ff9c8e94/types/node-fetch/index.d.ts#L46) type definition).

File: packages/integrations/node/test/api-route.test.js [hunk 4]
```diff
@@ -43,11 +46,47 @@ describe('API routes', () => {
...
+
 		handler(req, res);
-		req.send(Buffer.from(new Uint8Array([1, 2, 3, 4, 5])));
 
 		let [out] = await done;
 		let arr = Array.from(new Uint8Array(out.buffer));
 		expect(arr).to.deep.equal([5, 4, 3, 2, 1]);
 	});
+
+	it('Can post large binary daya', async () => {
```
Comment (matthewp) [inline packages/integrations/node/test/api-route.test.js:60]: typo
```suggestion
	it('Can post large binary data', async () => {
```

**Review by matthewp — APPROVED** [review]
lgtm!

**Review by bluwy — APPROVED** [review]
Looks great!

**Comment by hbgl** [comment]
I assume the failing check for **windows-latest (node@16) (pull_request)** is unrelated to the change. I just ran the tests locally on Windows and Node V16 and it passed.

**Comment by bluwy** [comment]
Yeah sorry about that. That particular test has been flaky. I re-ran it and will merge it once it passes.

**Pull request merged**